drivers/regulator/regulator_pca9420.c

```diff
@@ -322,76 +322,6 @@ static int regulator_pca9420_get_current_limit(const struct device *dev)
 	return MIN(config->max_ua, cconfig->vin_ilim_ua);
 }
 
-/*
```

Comment:
I don't think it makes sense to drop these mode specific configuration APIs. For example, the PCA9420 boots with different target voltages for each mode. So consider the following (toy) example:

- User designs system where SW1 output (defaults to 1.0V at powerup) is connected to a device that runs in the range 1.0-1.1V.
- User wants to utilize multiple modes in their application, and selects mode 1 as their second target mode
- Since SW1 defaults to an output voltage of 1.2V, the user effectively cannot switch to this mode.

I hope this makes it clear why I added mode support using these APIs initially. While this isn't a likely example, I still think the ability to configure target voltages before applying them is valuable to the user.

Reply:
Modes are defined as different operating modes by the regulator. For example, in case of nPM6001 you have hysteretic or PWM mode. Hysteretic is efficient at low load currents, whereas PWM is efficient at higher current loads.

![image](https://user-images.githubusercontent.com/25011557/205872098-02fe5cb4-6333-4d4c-a4bf-19d439e641a9.png)

So for example, in low-power mode you'll likely consume a few u/mA so hysteretic will give you the maximum efficiency and so battery life. When active, PWM mode will likely be the best choice. Linux even has an [API](https://elixir.bootlin.com/linux/latest/source/drivers/regulator/core.c#L4644) where you can specify your current load, so that the driver can choose the optimum mode internally.

From Linux regulator framework slides here: https://elinux.org/images/c/c1/Regulator-api-celf.pdf
![image](https://user-images.githubusercontent.com/25011557/205871327-b230120c-189e-42b6-b4de-b1cf6548eeae.png)

In case of PCA9420 I haven't seen any explanation of what modes 0...3 really mean. If they are just preprogrammed states, we may need to adjust the API for such functionality.

For reference, other examples:
https://www.ti.com/lit/ds/symlink/tps62866.pdf
https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/regulator/ti,tps62864.h

Reply:
Thank you for the clarification here. I was under the impression that modes were generally understood to be very hardware specific, and have different meanings across different regulators. If we want to use modes for operating efficiency at different loads, than the PCA9420 does not have modes in that sense.

The primary use case I am aware of for the modes present on the PCA9420 is automatically transitioning between groups of set voltages when changing power modes. On the RT685 and RT595, there are two output pins on the SOC that can be used to indicate the power state. The idea is that when paired with the PCA9420, these RT MCUs can seamlessly transition between power states, provided the voltages have been configured for each target mode:
![image](https://user-images.githubusercontent.com/34665051/205960982-54b46c1d-c82d-4958-9c45-532b87a52fa4.png)

If we want to use another API to configure these modes on the PCA9420, I'm open to that. Really, my primary concern is how to effectively enable this use case, which is required to optimize power consumption on these MCUs.

For more info on the power management optimization for RT595: https://www.nxp.com/docs/en/application-note/AN13162.pdf

Reply:
Thanks for all these details, I'll try to introduce some changes to the API to cover this case while keeping the usage of modes.

Reply:
Hi @gmarull, I've taken a look. I'd prefer that some standard API exists to cover the PCA9420's use case. While this is not a common feature, other regulators do exist that implement voltage scaling based on GPIO pin settings. Could we consider using `state` to describe the APIs present here? So these regulator APIs would be renamed to `regulator_set_state_voltage`, `regulator_set_state`, and `regulator_state_disable/enable`

Reply:
> Hi @gmarull, I've taken a look. I'd prefer that some standard API exists to cover the PCA9420's use case. While this is not a common feature, other regulators do exist that implement voltage scaling based on GPIO pin settings. Could we consider using `state` to describe the APIs present here? So these regulator APIs would be renamed to `regulator_set_state_voltage`, `regulator_set_state`, and `regulator_state_disable/enable`

OK, let me see what I can do. Note that the state in case of PCA9420 applies to all regulators, so I feel we need a separate API for the "parent regulator". Could you list other PMICs with this feature so I make sure what I propose solves more than a single usecase?